src/compiler/checker.ts

```diff
@@ -6638,6 +6686,412 @@ namespace ts {
+                    checkJsxExpression(<JsxExpression>node.children[i]);
+                }
+                else if (node.children[i].kind === SyntaxKind.JsxElement) {
+                    checkJsxElement(<JsxElement>node.children[i]);
+                }
+                else if (node.children[i].kind === SyntaxKind.JsxSelfClosingElement) {
+                    checkJsxSelfClosingElement(<JsxSelfClosingElement>node.children[i]);
+                }
+                else {
+                    // No checks for JSX Text
+                    Debug.assert(node.children[i].kind === SyntaxKind.JsxText);
+                }
+            }
+
+            return jsxElementType;
+        }
+
+        /// Returns true iff the JSX element name would be a valid identifier, ignoring restrictions about keywords not being identifiers
+        function isIdentifierLike(name: string) {
+            // - is the only character supported in JSX attribute names that isn't valid in JavaScript identifiers
+            return name.indexOf('-') < 0;
+        }
+
+        /// Returns true iff React would emit this tag name as a string rather than an identifier or qualified name
+        function isJsxIntrinsicIdentifier(tagName: Identifier|QualifiedName) {
+            if (tagName.kind === SyntaxKind.QualifiedName) {
+                return false;
+            }
+            else {
+                let firstChar = (<Identifier>tagName).text.charAt(0);
+                return firstChar.toLowerCase() === firstChar;
+            }
+        }
+
+        function checkJsxAttribute(node: JsxAttribute, elementAttributesType: Type, nameTable: Map<boolean>) {
+            var correspondingPropType: Type = undefined;
+
+            // Look up the corresponding property for this attribute
+            if (elementAttributesType === emptyObjectType && isIdentifierLike(node.name.text)) {
+                // If there is no 'props' property, you may not have non-"data-" attributes
+                error(node.parent, Diagnostics.JSX_element_class_does_not_support_attributes_because_it_does_not_have_a_0_property, getJsxElementPropertiesName());
+            }
+            else if (elementAttributesType && !isTypeAny(elementAttributesType)) {
+                var correspondingPropSymbol = getPropertyOfType(elementAttributesType, node.name.text);
+                correspondingPropType = correspondingPropSymbol && getTypeOfSymbol(correspondingPropSymbol);
+                if (!correspondingPropType) {
+                    // If there's no corresponding property with this name, error
+                    if (isIdentifierLike(node.name.text)) {
+                        error(node.name, Diagnostics.Property_0_does_not_exist_on_type_1, node.name.text, typeToString(elementAttributesType));
+                        return unknownType;
+                    }
+                }
+            }
+
+            let exprType: Type;
+            if (node.initializer) {

```

Comment:
1. It's more typical with the rest of our codebase, and consistency is important.
2. You can omit the type annotation.
3. You declare an uninitialized variable only to immediately initialize it in the following branches. This will sometimes signal that the variable has some complex initialization logic or may be undefined. Just initialize upon declaration.